src/jit/gentree.cpp

```diff
@@ -4203,14 +4203,21 @@ unsigned Compiler::gtSetEvalOrder(GenTree* tree)
             switch (op1Val->gtOper)
             {
                 case GT_IND:
+                case GT_BLK:
+                case GT_OBJ:
+                case GT_DYN_BLK:
 
-                    // Struct assignments are different from scalar assignments in that semantically
-                    // the address of op1 is evaluated prior to op2.
-                    if (!varTypeIsStruct(op1))
+                    // In an indirection, the destination address is evaluated prior to the source.
+                    // If we have any side effects on the target indirection,
+                    // we have to evaluate op1 first.
+                    // However, if op1 is a lclVar address, the side-effect flags don't really apply in
+                    // this situation (in fact, they are quite likely overly conservative; otherwise they
+                    // reflect the fact that an address-exposed local is being assigned to),
+                    // and SSA depends on having op2 evaluated first if the destination is a local.
```

Comment:
Hmm, I don't quite understand the comment. If an address-exposed local is being assigned to how is SSA involved in this?

Reply:
I guess I need to clarify the comment. What I meant to say was that if we have a lclVar assignment (i.e. the lhs is an indirection of an address involving a lclVar) that is marked with one of the `GTF_ALL_EFFECT` flags, either 1) it is simply overly conservative (this case, where we overly aggressively apply the `GLOB_REF` flag to `FIELD` nodes), or 2) it is address-taken. In either case, if we know that the lhs is a lclVar address, then it is safe to evaluate the rhs first, as the side effects on the lhs only apply to the actual assignment.

Reply:
> that is marked with one of the GTF_ALL_EFFECT flags, either 1) it is simply overly conservative (this case, where we overly aggressively apply the GLOB_REF flag to FIELD nodes), or 2) it is address-taken

Yes, I understand that, it's just the SSA bit that I find confusing. I guess that what you're saying is that it doesn't impact SSA, exactly because it's an address taken variable which SSA will just ignore.

Reply:
The code seems correct. Though I wish the JIT wouldn't do this dance of having the address of a lclvar being taken and then immediately assign to it through an indir. Such cases should be morphed to assign directly to the lclvar (of course, that wouldn't probably work for dynblk but that's a special and rare case).

Reply:
Yes, I *totally* agree. Would it be clearer if the comment said:
```
// However, if the LHS is a lclVar address, SSA relies on using evaluation order for its
// renaming, and therefore the RHS must be evaluated first.
// If we have an assignment involving a lclVar address, the LHS may be marked as having side-effects.
// However the side-effects won't require that we evaluate the LHS address first:
// - The GTF_GLOB_REF might have been conservatively set on a FIELD of a local.
// - The local might be address-exposed, but that side-effect happens at the actual assignment (not
//   when its address is "evaluated") so it doesn't change the side effect to "evaluate" the address
//   after the RHS (note that in this case it won't be renamed by SSA anyway, but the reordering is safe).
```